01-data-structures/04-hashes-part-1/hashclass.rb

```diff
@@ -1,27 +1,63 @@
+require_relative "hash_item"
+
 class HashClass
 
   def initialize(size)
     @items = Array.new(size)
   end
 
   def []=(key, value)
+    index = self.index(key, @items.length)
+
+    if @items[index].nil?
+      @items[index] = HashItem.new(key, value)
+    elsif @items[index].key == key
+      unless @items[index].value == value
+        @items[index].value = value
+        self.resize #for the spec
+      end
+    else
+      self.resize
+      self[key] = value
+    end
   end
 
-
   def [](key)
+    index = self.index(key, @items.length)
+
+    unless @items[index].nil? || @items[index].key != key
+      @items[index].value
+    end
   end
 
   def resize
+    old_array = @items
+    @items = Array.new(@items.length * 2)
+
+    old_array.each do |hash_item|
+      unless hash_item.nil?
+        index = self.index(hash_item.key, @items.length)
```

Comment:
Instead of putting the code from `[]=` in here, you could have just called `[]=` right away:

```ruby
old_array ...
  self[hash_item.key] = hash_item.value unless hash_item.nil?
end
```

There are some performance issues with that approach, but no more than many others.